codec/src/main/java/io/netty/handler/codec/ByteToMessageDecoder.java

```diff
@@ -351,6 +361,19 @@ protected void callDecode(ChannelHandlerContext ctx, ByteBuf in, List<Object> ou
                 if (isSingleDecode()) {
                     break;
                 }
+
+                if (out.size() < fireAfterReads) {
+                    continue;
+                }
+
+                if (out.size() == 1) {
+                    ctx.fireChannelRead(out.remove(0));
+                    continue;
+                }
+
+                for (int i = 0; i < out.size(); i++) {
+                    ctx.fireChannelRead(out.remove(0));
+                }
             }
```

Comment:
Use out.get(...) and call clear() on the end. remove(...) is expensive for an ArrayList